connector/apmconnector/transaction.go

```diff
@@ -0,0 +1,367 @@
+	SegmentNameProvider                     func(TransactionType) string
+	// FIXME
+	Span ptrace.Span
+}
+
+type TransactionsMap struct {
+	sqlParser    *SQLParser
+	apdex        Apdex
+	Transactions map[string]*Transaction
+}
+
+func NewTransactionsMap(apdexT float64) *TransactionsMap {
+	return &TransactionsMap{Transactions: make(map[string]*Transaction), sqlParser: NewSQLParser(), apdex: NewApdex(apdexT)}
+}
+
+func (transactions *TransactionsMap) ProcessTransactions() {
+	for _, transaction := range transactions.Transactions {
+		// if this returns false, we MAY not have seen all of the spans for a trace
+		transaction.ProcessRootSpan()
+	}
+}
+
+func (transactions *TransactionsMap) GetOrCreateTransaction(sdkLanguage string, span ptrace.Span, resourceMetrics *ResourceMetrics) (*Transaction, string) {
+	traceID := span.TraceID().String()
+	transaction, txExists := transactions.Transactions[traceID]
+	if !txExists {
+		transaction = &Transaction{SdkLanguage: sdkLanguage, SpanToChildDuration: make(map[string]int64),
+			resourceMetrics: resourceMetrics, Measurements: make(map[string]*Measurement), sqlParser: transactions.sqlParser, apdex: transactions.apdex}
+		transactions.Transactions[traceID] = transaction
+		//fmt.Printf("Created transaction for: %s   %s\n", traceID, transaction.sdkLanguage)
+	}
+
+	return transaction, traceID
+}
+
+func (transaction *Transaction) IsRootSet() bool {
+	return (ptrace.Span{}) != transaction.RootSpan
+}
+
+func (transaction *Transaction) SetRootSpan(span ptrace.Span) {
+	transaction.RootSpan = span
+}
+
+func (transaction *Transaction) AddSpan(span ptrace.Span) {
+	if span.Kind() == ptrace.SpanKindServer {
+		transaction.SetRootSpan(span)
+	} else {
+		isRoot := span.ParentSpanID().IsEmpty()
+		if isRoot {
+			transaction.SetRootSpan(span)
+		} else {
+			parentSpanID := span.ParentSpanID().String()
+			newDuration := DurationInNanos(span)
+
+			if measurement, exists := transaction.Measurements[parentSpanID]; exists {
+				measurement.ExclusiveDurationNanos -= newDuration
+			} else {
+				transaction.SpanToChildDuration[parentSpanID] += newDuration
+			}
+		}
+
+		if span.Kind() == ptrace.SpanKindClient {
+			// filter out db calls that have no parent (so no transaction)
+			if !isRoot {
+				transaction.ProcessClientSpan(span)
+			}
+		} else {
+			transaction.ProcessGenericSpan(span)
+		}
+	}
+}
+
+func NewSimpleNameProvider(name string) func(TransactionType) string {
+	return func(t TransactionType) string { return name }
+}
+
+func (transaction *Transaction) AddMeasurement(measurement *Measurement) {
+	transaction.Measurements[measurement.SpanID] = measurement
+	measurement.ExclusiveDurationNanos = measurement.ExclusiveTime(transaction)
+	measurement.Attributes.PutStr("metricTimesliceName", measurement.MetricTimesliceName)
+}
+
+func (transaction *Transaction) ProcessDatabaseSpan(span ptrace.Span) bool {
+	if dbSystem, dbSystemPresent := span.Attributes().Get(DbSystemAttributeName); dbSystemPresent {
```

Comment:
[nitpick] early returns here for `dbSystemPresent` and `dbOperationPresent` would remove nested ifs